File: java/src/main/java/com/google/cloud/dataproc/templates/gcs/GCSToJDBC.java
/*
 * Copyright (C) 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.gcs;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_AVRO_FORMAT;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_CSV_FORMAT;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_CSV_HEADER;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_CSV_INFER_SCHEMA;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_ORC_FORMAT;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.GCS_JDBC_PRQT_FORMAT;

import com.google.cloud.dataproc.dialects.SpannerJdbcDialect;
import com.google.cloud.dataproc.templates.BaseTemplate;
import com.google.cloud.dataproc.templates.util.PropertyUtil;
import com.google.cloud.dataproc.templates.util.ValidationUtil;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions;
import org.apache.spark.sql.jdbc.JdbcDialects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GCSToJDBC implements BaseTemplate {

  private static final Logger LOGGER = LoggerFactory.getLogger(GCSToJDBC.class);
  private final GCSToJDBCConfig config;

  public GCSToJDBC(GCSToJDBCConfig config) {
    this.config = config;
  }

  public static GCSToJDBC of(String... args) {
    GCSToJDBCConfig config = GCSToJDBCConfig.fromProperties(PropertyUtil.getProperties());
    ValidationUtil.validateOrThrow(config);
    LOGGER.info("Config loaded\n{}", config);
    return new GCSToJDBC(config);
  }

  @Override
  public void runTemplate() {
    validateInput();
    try (SparkSession spark = SparkSession.builder().appName("GCS to JDBC").getOrCreate()) {
      // Set log level
      spark.sparkContext().setLogLevel(config.getSparkLogLevel());

      Dataset<Row> dataset = load(spark);

      write(dataset);
    }
  }

  public Dataset<Row> load(SparkSession spark) {
    Dataset<Row> inputData = null;

    switch (config.getInputFormat()) {
      case GCS_JDBC_CSV_FORMAT:
        inputData =
            spark
                .read()
                .format(config.getInputFormat())
                .option(GCS_JDBC_CSV_HEADER, true)
                .option(GCS_JDBC_CSV_INFER_SCHEMA, true)
                .load(config.getInputLocation());
        break;
      case GCS_JDBC_AVRO_FORMAT:
      case GCS_JDBC_ORC_FORMAT:
        inputData = spark.read().format(config.getInputFormat()).load(config.getInputLocation());
        break;
      case GCS_JDBC_PRQT_FORMAT:
        inputData = spark.read().parquet(config.getInputLocation());
        break;
      default:
        throw new IllegalArgumentException(
            "Currently avro, orc, parquet and csv are the only supported formats");
    }

    return inputData;
  }

  public void write(Dataset<Row> dataset) {
    JdbcDialects.registerDialect(new SpannerJdbcDialect());

    int current_partitions = dataset.toJavaRDD().getNumPartitions();
    if (config.getCustomSparkPartitions() == null || config.getCustomSparkPartitions().isEmpty()) {
      int custom_partitions = current_partitions;
    } else {
      int custom_partitions = Integer.parseInt(config.getCustomSparkPartitions());
      LOGGER.info("Current Spark partitions: ", Integer.toString(current_partitions));

      if (current_partitions > custom_partitions) {
        LOGGER.info(
            "Current partitions are greater than provided custom partitions. Will use coalesce()... ");
        dataset = dataset.coalesce(custom_partitions);
      } else if (current_partitions < custom_partitions) {
        LOGGER.info(
            "Current partitions are smaller than provided custom partitions. Will use repartition()... ");
        dataset = dataset.repartition(custom_partitions);
      } else {
        LOGGER.info("Current partitions are equal to provided custom partitions. ");
      }
    }

    dataset
        .write()
        .format("jdbc")
        .option(JDBCOptions.JDBC_URL(), config.getJDBCUrl())
        .option(JDBCOptions.JDBC_TABLE_NAME(), config.getTable())
        .option(
            JDBCOptions.JDBC_BATCH_INSERT_SIZE(), config.getBatchInsertSize()) // default is 1000
        .option(JDBCOptions.JDBC_DRIVER_CLASS(), config.getJDBCDriver())
        .mode(config.getSaveMode())
        .save();
  }

  public void validateInput() {}
}

File: python/dataproc_templates/gcs/gcs_to_jdbc.py
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint

from pyspark.sql import SparkSession

from dataproc_templates import BaseTemplate
import dataproc_templates.util.template_constants as constants
from dataproc_templates.util.argument_parsing import add_spark_options
from dataproc_templates.util.dataframe_reader_wrappers import ingest_dataframe_from_cloud_storage


__all__ = ['GCSToJDBCTemplate']


class GCSToJDBCTemplate(BaseTemplate):
    """
    Dataproc template implementing loads from Cloud Storage into JDBC
    """

    @staticmethod
    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
        parser: argparse.ArgumentParser = argparse.ArgumentParser()

        parser.add_argument(
            f'--{constants.GCS_JDBC_INPUT_LOCATION}',
            dest=constants.GCS_JDBC_INPUT_LOCATION,
            required=True,
            help='Cloud Storage location of the input files'
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_INPUT_FORMAT}',
            dest=constants.GCS_JDBC_INPUT_FORMAT,
            required=True,
            help='Input file format (one of: avro,parquet,csv,json,delta)',
            choices=[
                constants.FORMAT_AVRO,
                constants.FORMAT_PRQT,
                constants.FORMAT_CSV,
                constants.FORMAT_JSON,
                constants.FORMAT_DELTA
            ]
        )
        add_spark_options(parser, constants.get_csv_input_spark_options("gcs.jdbc.input."))
        parser.add_argument(
            f'--{constants.GCS_JDBC_OUTPUT_TABLE}',
            dest=constants.GCS_JDBC_OUTPUT_TABLE,
            required=True,
            help='JDBC output table name'
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_OUTPUT_MODE}',
            dest=constants.GCS_JDBC_OUTPUT_MODE,
            required=False,
            default=constants.OUTPUT_MODE_APPEND,
            help=(
                'Output write mode '
                '(one of: append,overwrite,ignore,errorifexists) '
                '(Defaults to append)'
            ),
            choices=[
                constants.OUTPUT_MODE_OVERWRITE,
                constants.OUTPUT_MODE_APPEND,
                constants.OUTPUT_MODE_IGNORE,
                constants.OUTPUT_MODE_ERRORIFEXISTS
            ]
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_OUTPUT_URL}',
            dest=constants.GCS_JDBC_OUTPUT_URL,
            required=True,
            help='JDBC output URL'
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_OUTPUT_DRIVER}',
            dest=constants.GCS_JDBC_OUTPUT_DRIVER,
            required=True,
            help='JDBC output driver name'
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_BATCH_SIZE}',
            dest=constants.GCS_JDBC_BATCH_SIZE,
            required=False,
            default=1000,
            help='JDBC output batch size'
        )
        parser.add_argument(
            f'--{constants.GCS_JDBC_NUMPARTITIONS}',
            dest=constants.GCS_JDBC_NUMPARTITIONS,
            required=False,
            help='The maximum number of partitions to be used for parallelism in table writing'
        )

        known_args: argparse.Namespace
        known_args, _ = parser.parse_known_args(args)

        return vars(known_args)

    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:
        logger: Logger = self.get_logger(spark=spark)

        # Arguments
        input_location: str = args[constants.GCS_JDBC_INPUT_LOCATION]
        input_format: str = args[constants.GCS_JDBC_INPUT_FORMAT]
        jdbc_url: str = args[constants.GCS_JDBC_OUTPUT_URL]
        jdbc_table: str = args[constants.GCS_JDBC_OUTPUT_TABLE]
        output_mode: str = args[constants.GCS_JDBC_OUTPUT_MODE]
        output_driver: str = args[constants.GCS_JDBC_OUTPUT_DRIVER]
        batch_size: int = args[constants.GCS_JDBC_BATCH_SIZE]
        jdbc_numpartitions: int = args[constants.GCS_JDBC_NUMPARTITIONS]

        ignore_keys = {constants.GCS_JDBC_OUTPUT_URL}
        filtered_args = {key:val for key,val in args.items() if key not in ignore_keys}
        logger.info(
            "Starting Cloud Storage to JDBC Spark job with parameters:\n"
            f"{pprint.pformat(filtered_args)}"
        )

        # Read
        input_data = ingest_dataframe_from_cloud_storage(spark, args, input_location, input_format, "gcs.jdbc.input.")

        # Write
        if not jdbc_numpartitions:
            jdbc_numpartitions = input_data.rdd.getNumPartitions()

        # TODO Convert this call to a function in dataproc_templates.util.dataframe_writer_wrappers
        input_data.write \
            .format(constants.FORMAT_JDBC) \
            .option(constants.JDBC_URL, jdbc_url) \
            .option(constants.JDBC_TABLE, jdbc_table) \
            .option(constants.JDBC_DRIVER, output_driver) \
            .option(constants.JDBC_BATCH_SIZE, batch_size) \
            .option(constants.JDBC_NUMPARTITIONS, jdbc_numpartitions) \
            .mode(output_mode) \
            .save()
